okhttp/src/main/java/okhttp3/Authenticator.java

```diff
@@ -52,11 +52,33 @@
     @Override public Request authenticate(Route route, Response response) {
       return null;
     }
+
+    @Override
+    public Request authenticatePreemptive(Route route, Request request) throws IOException {
+      return null;
+    }
+
+    @Override
+    public boolean isPreemptive() {
+      return false;
+    }
   };
 
   /**
    * Returns a request that includes a credential to satisfy an authentication challenge in {@code
    * response}. Returns null if the challenge cannot be satisfied.
    */
   Request authenticate(Route route, Response response) throws IOException;
+
+  /**
+   * Return a request that is a copy of the original request plus authentication header.
+   * isPreemptive method must return true.
+   *
+   */
+  Request authenticatePreemptive(Route route, Request request) throws IOException;
```

Comment:
These two methods will break binary compatibility for existing implementors. Since OkHttp only allows a single Authenticator, but may be servicing multiple hosts concurrently, maybe the best compromise is an extended interface?